### PR TITLE
Fix Java 18.3 compile/build errors

### PR DESCRIPTION
**Pull request by JasonFengJ9**
Fix Java 18.3 compile/build errors

1. Added `#define J2SE_2018_3  0xB700` for Java 18.3;
2. Added definitions for `J2SE_LATEST` and `J2SE_SHAPE_LATEST` according to flag `JAVA_SPEC_VERSION`;
3. Added function to get JVM SE Version info from release file when `classlib.properties` doesn't present;
4. If no version info found from either `classlib.properties` or `release` file, returns `J2SE_LATEST | J2SE_SHAPE_LATEST`.

Closes: #293

Reviewer @pshipton 
FYI: @DanHeidinga @adamfarley 

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

### CODE REVIEW

File: runtime/j9vm/jvm.c [hunk 1]
```diff
@@ -1218,73 +1226,148 @@ decodeSetting(const char* key, const char* value, VersionSetting* settings, IDAT
...
- * classpath based on data found there.
- * @param vm
- * @param cursor
- * @return TRUE on success, FALSE otherwise.
+ * Get Java version info from 'classlib.properties' file
+ * Attempt loading 'classlib.properties' file, if found, 'shape' & 'version' values are retrieved,
+ * and decoded as J2SE_xx and J2SE_SHAPE_xx accordingly. 
+ * J2SE_xx | J2SE_SHAPE_xx will be returned.
+ *
+ * @return jvm SE Version value
```
Comment (pshipton) [inline runtime/j9vm/jvm.c]: Please include the behavior when classlib.properties isn't found.

File: runtime/oti/j2sever.h [hunk 2]
```diff
@@ -27,24 +27,17 @@
...
+ * 		 This allows JVM operates with latest version when neither classlib.properties
+ * 		 nor release file presents.
  */
 #define J2SE_15  0x1500
 #define J2SE_16  0x1600
 #define J2SE_17  0x1700
 #define J2SE_18  0x1800
 #define J2SE_19  0x1900
-#define J2SE_LATEST  J2SE_19
+#define J2SE_B7  0xB700				/* 0xB7 is 183 which refers to Java 18.3 */
```
Comment (pshipton) [inline runtime/oti/j2sever.h]: You can use B7 for the value, but the #define needs a better name, like J2SE_18_3

File: runtime/j9vm/jvm.c [hunk 3]
```diff
@@ -1218,73 +1226,148 @@ decodeSetting(const char* key, const char* value, VersionSetting* settings, IDAT
...
+	return finalVersion;
+}
+
+/**
+ * Get Java version info from 'release' file
+ * Attempt loading 'release' file, if found, 'JAVA_VERSION' value is retrieved, and decoded as following:
+ * "1.8.0_xxx"	--- Java 8 (J2SE_18 | J2SE_SHAPE_SUN)
+ * "9"			--- Java 9 (J2SE_19 | J2SE_SHAPE_B165)
+ * "10"			--- Java 18.3 (J2SE_B7 | J2SE_SHAPE_B1803)
+ * Others		--- (ßJ2SE_LATEST | J2SE_SHAPE_LATEST)
```
Comment (pshipton) [inline runtime/j9vm/jvm.c]: There appears to be a non-ascii character before "J2SE_LATEST"

File: runtime/j9vm/jvm.c [hunk 4]
```diff
@@ -1218,73 +1226,148 @@ decodeSetting(const char* key, const char* value, VersionSetting* settings, IDAT
...
+ * Note: 'release' file contains JAVA_VERSION="10" for Java 18.3 at this moment.
+ *
+ * @return jvm SE Version value
```
Comment (pshipton) [inline runtime/j9vm/jvm.c]: Please include the behavior when the release file isn't found.

File: runtime/j9vm/jvm.c [hunk 5]
```diff
@@ -1218,73 +1226,148 @@ decodeSetting(const char* key, const char* value, VersionSetting* settings, IDAT
...
 /**
- * Loads the classlib.properties file and initialize the bootstrap
- * classpath based on data found there.
- * @param vm
- * @param cursor
- * @return TRUE on success, FALSE otherwise.
+ * Get Java version info from 'classlib.properties' file
+ * Attempt loading 'classlib.properties' file, if found, 'shape' & 'version' values are retrieved,
+ * and decoded as J2SE_xx and J2SE_SHAPE_xx accordingly. 
+ * J2SE_xx | J2SE_SHAPE_xx will be returned.
```
Comment (pshipton) [inline runtime/j9vm/jvm.c]: This is a better description for @return, but all the behavior when classlib.properties isn't found.

File: runtime/j9vm/jvm.c (hunk 4 above)
Comment (pshipton) [reply runtime/j9vm/jvm.c]: Please update the @return as per the comments on getVersionFromClasslibPropertiesFile().

File: runtime/j9vm/jvm.c [hunk 6]
```diff
@@ -1218,73 +1226,148 @@ decodeSetting(const char* key, const char* value, VersionSetting* settings, IDAT
...
+
+	return finalVersion;
+}
+
+/**
+ * Get Java version of running JVM
+ * Attempt loading the 'classlib.properties' file first, otherwise try 'release' file,
+ * if neither of files is found, J2SE_LATEST | J2SE_SHAPE_LATEST will be returned.
+ *
+ * @return jvmSEVersion value
```
Comment (pshipton) [inline runtime/j9vm/jvm.c]: Please update the @return as per the comments on getVersionFromClasslibPropertiesFile(). jvmSEVersion is an internal implementation detail.

File: jcl/src/java.base/share/classes/java/lang/invoke/LambdaForm.java [hunk 7]
```diff
@@ -116,7 +116,13 @@ static BasicType basicType(Class<?> cls) {
...
 	/*[IF Java18.3]*/
 	enum Kind {
-		PLACEHOLDER;
+		CONVERT,
+		SPREAD,
+		COLLECT,
+		GUARD,
+		GUARD_WITH_CATCH,
+		LOOP,
+		TRY_FINALLY
```
Comment (pshipton) [inline jcl/src/java.base/share/classes/java/lang/invoke/LambdaForm.java]: If this change was a separate Pull Request I would have merged it. It doesn't need to be mixed in with the unrelated changes to determine the java version.

Comment (JasonFengJ9) [reply jcl/src/java.base/share/classes/java/lang/invoke/LambdaForm.java]: This `j.l.i.LambdaForm.Kind` enum change alone can't fix `Java 18.3` compile/build errors. 
To fix `Java 18.3` compile/build errors, this `j.l.i.LambdaForm.Kind` enum change need to be applied with other part of this pull request. Agreed, this change can be in a separate PR for quick reviewing/merging. Created https://github.com/eclipse/openj9/pull/377.

File: runtime/oti/j2sever.h (hunk 2 above)
Comment (JasonFengJ9) [reply runtime/oti/j2sever.h]: `J2SE_18_3` was my initial choice, just felt it is easy to confuse with `J2SE_18`.

**Comment by JasonFengJ9** [comment]
Pull request has been updated, pls review again.

File: runtime/oti/j2sever.h (hunk 2 above)
Comment (pshipton) [reply runtime/oti/j2sever.h]: We can brainstorm to find a better name. 
J2SE_2018_3 ?

Comment (JasonFengJ9) [reply runtime/oti/j2sever.h]: How about `JAVA_18_3` which differentiate from earlier name conversion `J2SE_xx`?

Comment (pshipton) [reply runtime/oti/j2sever.h]: Then its not or less obvious that JAVA_* and J2SE_* are related constants that are used interchangeably. Since 18.3 is short form for March 2018, I figure using 2018 would be clear and could not be confused with 18 meaning Java 8.

Comment (dmitripivkine) [reply runtime/oti/j2sever.h]: We are using this definition to differentiate implementations in the code like "if (J2SE_VERSION((vm) >= J2SE_19)". J2SE_B7 would be larger then any of them (the check uses unsigned, does not it?) including 1.9.

Comment (JasonFengJ9) [reply runtime/oti/j2sever.h]: `J2SE_2018_3` works for me.

Comment (JasonFengJ9) [reply runtime/oti/j2sever.h]: @dmitripivkine right, that's the purpose, `Java 18.3 (0xB700)` should be superset of `Java 9 (0x1900)`.

Comment (dmitripivkine) [reply runtime/oti/j2sever.h]: What we are going to do with Java 10 (0x1A00 I assume) ?

Comment (JasonFengJ9) [reply runtime/oti/j2sever.h]: `Java 18.3` is `Java 10` (the next version we thought). 
https://blogs.oracle.com/java-platform-group/faster-and-easier-use-and-redistribution-of-java-se

File: runtime/j9vm/jvm.c [hunk 8]
```diff
@@ -1218,73 +1226,149 @@ decodeSetting(const char* key, const char* value, VersionSetting* settings, IDAT
...
- * classpath based on data found there.
- * @param vm
- * @param cursor
- * @return TRUE on success, FALSE otherwise.
+ * Attempt loading 'classlib.properties' file, and get Java version info.
+ * If found, 'version' and 'shape' values are retrieved, and decoded as J2SE_xx and J2SE_SHAPE_xx accordingly.
+ * 'J2SE_xx | J2SE_SHAPE_xx' will be returned;
+ * If not found, 0 is returned.
+ *
+ * @return jvm SE Version value
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply runtime/j9vm/jvm.c]: Changeset has been updated, pls have another look.

**Comment by JasonFengJ9** [comment]
Updated the changeset to get `J2SE_LATEST` & `J2SE_SHAPE_LATEST` values for different Java version as per compilation flag `JAVA_SPEC_VERSION` to minimize OpenJDK JCL patches. 
Pls review.

**Comment by JasonFengJ9** [comment]
Updated the changeset to accept `VERSION_MAJOR` from makefile, and default to `9` if it is not defined.
Pls review.

File: runtime/makelib/mkconstants.mk.ftl [hunk 9]
```diff
@@ -24,7 +24,10 @@
 </#list>
 
 # Define the Java Version we are compiling
+# Ensure VERSION_MAJOR hence JAVA_SPEC_VERSION not empty
+ifndef VERSION_MAJOR
 VERSION_MAJOR:=9
```
Comment (keithc-ca) [inline runtime/makelib/mkconstants.mk.ftl]: This can be more simply expressed as `VERSION_MAJOR ?= 9`; there's no need for `ifndef`.

Comment (JasonFengJ9) [reply runtime/makelib/mkconstants.mk.ftl]: Updated with comments.

File: runtime/makelib/mkconstants.mk.ftl [hunk 10]
```diff
@@ -24,10 +24,12 @@
 </#list>
 
 # Define the Java Version we are compiling
+# This VERSION_MAJOR variable will be overridden by command line option 
+# when there is one such as VERSION_MAJOR=xx.
```
Comment (keithc-ca) [inline runtime/makelib/mkconstants.mk.ftl]: I don't think `VERSION_MAJOR` will always be defined on the command line (in all recursive calls to make), so the definition needs to use `?=`, not `:=`.

Comment (keithc-ca) [reply runtime/makelib/mkconstants.mk.ftl]: I have been convinced that `:=` is fine (perhaps even preferred).

File: runtime/j9vm/jvm.c [hunk 11]
```diff
@@ -1218,73 +1226,156 @@ decodeSetting(const char* key, const char* value, VersionSetting* settings, IDAT
...
+	propsPathBuffer = jvmBufferCat(propsPathBuffer, jvmBufferData(j9Buffer));
+	propsPathBuffer = jvmBufferCat(propsPathBuffer, DIR_SEPARATOR_STR "release");
+	propsFile = props_file_open(PORTLIB, jvmBufferData(propsPathBuffer), NULL, 0);
+	free(propsPathBuffer);
+	propsPathBuffer = NULL;
+	if (NULL != propsFile) {
+		const char *version = props_file_get(propsFile, "JAVA_VERSION");
+		if (NULL != version) {
+#define		JAVA_VERSION_8		"\"1.8.0"		/* its usual format is "1.8.0_xxx" */
+			if (!strncmp(version, JAVA_VERSION_8, strlen(JAVA_VERSION_8))) {
```
Comment (keithc-ca) [inline runtime/j9vm/jvm.c]: Using `sizeof(JAVA_VERSION_8) - 1` avoids the runtime function call to `strlen`.

Comment (JasonFengJ9) [reply runtime/j9vm/jvm.c]: Fixed. Pls have another look.
